### ISSUE TITLE
Annulus particle advection error

### ISSUE THREAD

**Issue by rsbyrne** [underworldcode/underworld2#372]
I've been having this issue for some time. It occurs at seemingly random times whenever a swarm is being advected in the annulus. It probably relates to particles being out of bounds. Reloading from a previous checkpoint allows one to continue past the point where the error occurred, making me believe that it can't be a problem with the model.

Attached is a zip file containing a model I ran back in March that reproduces the error (accessed by running 'localscript.py') and outputs of the velocity field and swarm from the time step where the error occurred. Pasted below is the error message from that encounter. Note that this model has an omission (it should have a '+ 0. * velocityField[0] term in the viscosityFn that flags to the solver that the problem is nonlinear) - however, the model I ran just today is the same in every way except that it does not have the omission, and it hit exactly the same bug, suggesting that this is a separate issue.

Also of note is that I have run this same model on my local machine (a Linux box that sits on my desk) and have had no problems. Yet when running the model on a cluster machine (in this case, a sixteen-core Nectar virtual machine) the problem mysteriously reappears. I am running in parallel.

I will endeavour to catch a fresh example of the bug for all to peruse.

Cheers

[debug.zip](https://github.com/underworldcode/underworld2/files/3097231/debug.zip)

```
SCR Solver Summary:

  Multigrid setup:        = 0.003017 secs 
  RHS V Solve:            = 0.01836 secs / 22 its
  Pressure Solve:         = 0.4213 secs / 30 its
  Final V Solve:          = 0.01627 secs / 19 its

  Total BSSCR Linear solve time: 0.467418 seconds

Linear solver (NISMZXXW__system-execute), solution time 4.685843e-01 (secs)
Linear solver (3M3K7GGB__system-execute) 
Linear solver (3M3K7GGB__system-execute), solution time 2.779317e-02 (secs)
Time Integration
An uncaught exception was encountered on processor 1.
Traceback (most recent call last):
  File "localscript.py", line 12, in <module>
    system.iterate()
  File "/workspace/user_data/MS98X_debug.py", line 279, in iterate
    dt = integrate()
  File "/workspace/user_data/MS98X_debug.py", line 273, in integrate
    advector.integrate(dt)
  File "/opt/underworld2/underworld/timing.py", line 323, in timed
    return routine(*args, **kwargs)
  File "/opt/underworld2/underworld/systems/_timeintegration.py", line 152, in integrate
    super(SwarmAdvector,self).integrate(dt)
  File "/opt/underworld2/underworld/timing.py", line 323, in timed
    return routine(*args, **kwargs)
  File "/opt/underworld2/underworld/systems/_timeintegration.py", line 50, in integrate
    libUnderworld.StgDomain._TimeIntegrator_Execute(self._cself, None)
RuntimeError: Error encountered. Full restart recommended as exception safety not guaranteed. Error message:
Error - in TimeIntegrand_SecondOrder(), for TimeIntegrand "CRQ0U9F5__integrand" of type SwarmAdvector: When trying to find time deriv for item 15481 in step 1, *failed*.

application called MPI_Abort(comm=0x84000002, 1) - process 1
```

**Comment by jmansour** [comment]
Yeah looks like particles out of bound. Perhaps we can made a further outer
layer of pseudo elements, with new outer nodes taking theirs values from
their radial direct neighbour.

I'm not sure what the implications would be for particle populations near
the boundaries, but it might tell us something about the advection scheme
in any case.

On 19 Apr. 2019 4:45 pm, "Rohan S Byrne" <notifications@github.com> wrote:

I've been having this issue for some time. It occurs at seemingly random
times whenever a swarm is being advected in the annulus. It probably
relates to particles being out of bounds. Reloading from a previous
checkpoint allows one to continue past the point where the error occurred,
making me believe that it can't be a problem with the model.

Here is the error message. Also attached is a zip file containing a model
that reproduces the error (accessed by running 'localscript.py') and
outputs of the velocity field and swarm from the time step where the error
occurred.

Cheers
debug.zip
<https://github.com/underworldcode/underworld2/files/3097231/debug.zip>

SCR Solver Summary:

  Multigrid setup:        = 0.003017 secs
  RHS V Solve:            = 0.01836 secs / 22 its
  Pressure Solve:         = 0.4213 secs / 30 its
  Final V Solve:          = 0.01627 secs / 19 its

  Total BSSCR Linear solve time: 0.467418 seconds

Linear solver (NISMZXXW__system-execute), solution time 4.685843e-01 (secs)
Linear solver (3M3K7GGB__system-execute)
Linear solver (3M3K7GGB__system-execute), solution time 2.779317e-02 (secs)
Time Integration
An uncaught exception was encountered on processor 1.
Traceback (most recent call last):
  File "localscript.py", line 12, in <module>
    system.iterate()
  File "/workspace/user_data/MS98X_debug.py", line 279, in iterate
    dt = integrate()
  File "/workspace/user_data/MS98X_debug.py", line 273, in integrate
    advector.integrate(dt)
  File "/opt/underworld2/underworld/timing.py", line 323, in timed
    return routine(*args, **kwargs)
  File "/opt/underworld2/underworld/systems/_timeintegration.py", line
152, in integrate
    super(SwarmAdvector,self).integrate(dt)
  File "/opt/underworld2/underworld/timing.py", line 323, in timed
    return routine(*args, **kwargs)
  File "/opt/underworld2/underworld/systems/_timeintegration.py", line
50, in integrate
    libUnderworld.StgDomain._TimeIntegrator_Execute(self._cself, None)
RuntimeError: Error encountered. Full restart recommended as exception
safety not guaranteed. Error message:
Error - in TimeIntegrand_SecondOrder(), for TimeIntegrand
"CRQ0U9F5__integrand" of type SwarmAdvector: When trying to find time
deriv for item 15481 in step 1, *failed*.

application called MPI_Abort(comm=0x84000002, 1) - process 1

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/underworldcode/underworld2/issues/372>, or mute the
thread

**Comment by julesghub** [comment]
Unfortunately I can't reproduce error on 1 or 4 procs. @rsbyrne how many procs did you run on?

**Comment by rsbyrne** [comment]
Update on this:

Looks like, whatever was causing the problem, it no longer manifests on the most up to date image. Closing this issue but will keep an eye out for anything like it going forward.
